managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorContainer.java

```diff
@@ -204,6 +281,24 @@ public ManagedCursor getSlowestReader() {
         }
     }
 
+    /**
+     * @return Returns the CursorInfo for the cursor with the oldest position,
+     *         or null if there aren't any tracked cursors
+     */
+    public CursorInfo getCursorWithOldestPosition() {
+        long stamp = rwLock.readLock();
+        try {
+            if (heap.isEmpty()) {
+                return null;
+            } else {
+                Item item = heap.get(0);
```

Comment:
I know it's old codes, but still, I don't understand why don't we use PriorityQueue to sort cursors.

Reply:
Fair question. @merlimat made the initial import in 2016, so there is a chance he'll know.

Reply:
@dragosvictor Actually, we can use `PriorityQueue#remove(Object o)` or `PriorityQueue#removeIf(Predicate<Object> filter)` to remove elements randomly. But it doesn't matter, the current implementation also works fine.

Reply:
@dragosvictor answered [here](https://github.com/apache/pulsar/pull/21816#discussion_r1437844593). It's because removing an arbitrary object from a PriorityQueue is O(n) while in Pulsar implementation, removing is O(1) due to the usage of Map of cursor name to `Item` and in it, we have its index in the array.

Reply:
I thought that also, but when you remove, you still sort, so it is still O(N) like queue remove complexity.

Reply:
For a removal, it would be O(logN): 
 1. Removing from skip-list is logN
 2. Removing from heap is also logN since we don't need to sort it